internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,479 @@
+type SinglePing struct {
+	ConnIdDst   string                         `json:"conn_id_dst"`
+	ConnIdSrc   string                         `json:"conn_id_src"`
+	Failure     *string                        `json:"failure"`
+	Request     *model.ArchivalMaybeBinaryData `json:"request"`
+	RequestTime string                         `json:"request_time"`
+	Responses   []*SinglePingResponse          `json:"responses"`
+}
+
+type SinglePingResponse struct {
+	Data              *model.ArchivalMaybeBinaryData `json:"response_data"`
+	Failure           *string                        `json:"failure"`
+	ResponseTime      string                         `json:"response_time"`
+	SupportedVersions []uint32                       `json:"supported_versions"`
+}
+
+// makeResponse is a utility function to create a SinglePingResponse
+func makeResponse(resp *responseInfo) *SinglePingResponse {
+	var data *model.ArchivalMaybeBinaryData
+	if resp.raw != nil {
+		data = &model.ArchivalMaybeBinaryData{Value: string(resp.raw)}
+	}
+	return &SinglePingResponse{
+		Data:              data,
+		Failure:           archival.NewFailure(resp.err),
+		ResponseTime:      formatTime(resp.t),
+		SupportedVersions: resp.versions,
+	}
+}
+
+// Measurer performs the measurement.
+type Measurer struct {
+	config Config
+}
+
+// ExperimentName implements ExperimentMeasurer.ExperimentName.
+func (m *Measurer) ExperimentName() string {
+	return testName
+}
+
+// ExperimentVersion implements ExperimentMeasurer.ExperimentVersion.
+func (m *Measurer) ExperimentVersion() string {
+	return testVersion
+}
+
+// pingInfo contains information about a ping request
+// and the corresponding ping responses
+type pingInfo struct {
+	request   *requestInfo
+	responses []*responseInfo
+}
+
+// requestInfo contains the information of a sent ping request.
+type requestInfo struct {
+	t     time.Time
+	raw   []byte
+	dstID string
+	srcID string
+	err   error
+}
+
+// responseInfo contains the information of a received ping reponse.
+type responseInfo struct {
+	t        time.Time
+	raw      []byte
+	dstID    string
+	versions []uint32
+	err      error
+}
+
+// sender sends a ping requests to the target hosts every second
+func (m *Measurer) sender(
+	ctx context.Context,
+	pconn model.UDPLikeConn,
+	destAddr *net.UDPAddr,
+	out chan<- requestInfo,
+	sess model.ExperimentSession,
+) {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	i := 0
+	for i < int(m.config.repetitions()) {
+		select {
+		case <-ctx.Done():
+			return // user aborted or timeout expired
+
+		case sendTime := <-ticker.C:
+			packet, dstID, srcID := buildPacket()     // build QUIC Initial packet
+			_, err := pconn.WriteTo(packet, destAddr) // send Initial packet
+
+			sess.Logger().Infof("PING %s", destAddr)
+
+			// propagate send information, including errors
+			out <- requestInfo{raw: packet, t: sendTime, dstID: hex.EncodeToString(dstID), srcID: hex.EncodeToString(srcID), err: err}
+			i += 1
```

Comment:
```suggestion
```

Line not needed if we use the `for` notation.

Reply:
Also, I think here we should handle the `errors.Is(err, net.ErrClosed)` case by returning

Reply:
*correction*: we should handle this case before propagating the error: I guess we don't want to propagate it?